From 012b9fcc949cd85434adf3d9fc12ec65c2e5c57f Mon Sep 17 00:00:00 2001
From: kadarlax <raghuveerx.kadarla@intel.com>
Date: Mon, 6 Sep 2021 19:33:55 +0530
Subject: [PATCH 094/109] misc: xlink-pcie: API to allocate DMA memory

Add API's to allocate dma_alloc_coherent memory needed by,
flash logic module.

Signed-off-by: kadarlax <raghuveerx.kadarla@intel.com>
---
 drivers/misc/xlink-pcie/remote_host/pci.c | 13 +++++++++++++
 drivers/misc/xlink-pcie/remote_host/pci.h |  5 +++++
 2 files changed, 18 insertions(+)

diff --git a/drivers/misc/xlink-pcie/remote_host/pci.c b/drivers/misc/xlink-pcie/remote_host/pci.c
index 4a6bfe9f0833..d3d847acda49 100644
--- a/drivers/misc/xlink-pcie/remote_host/pci.c
+++ b/drivers/misc/xlink-pcie/remote_host/pci.c
@@ -313,6 +313,11 @@ int intel_xpcie_pci_init(struct xpcie_dev *xdev, struct pci_dev *pdev)
 	if (!rc)
 		goto init_exit;
 
+#if (IS_ENABLED(CONFIG_ARCH_THUNDERBAY))
+	xdev->fl_vbuf = NULL;
+	xdev->fl_buf_size = 0;
+#endif
+
 error_dma_mask:
 	intel_xpcie_pci_unmap_bar(xdev);
 
@@ -349,6 +354,14 @@ int intel_xpcie_pci_cleanup(struct xpcie_dev *xdev)
 	xdev->core_irq_callback = NULL;
 	intel_xpcie_pci_irq_cleanup(xdev);
 
+#if (IS_ENABLED(CONFIG_ARCH_THUNDERBAY))
+	if (xdev->fl_vbuf) {
+		dma_free_coherent(&xdev->pci->dev, xdev->fl_buf_size,
+				  xdev->fl_vbuf, xdev->fl_phys_addr);
+		xdev->fl_vbuf = NULL;
+		xdev->fl_buf_size = 0;
+	}
+#endif
 	intel_xpcie_core_cleanup(&xdev->xpcie);
 
 	intel_xpcie_pci_unmap_bar(xdev);
diff --git a/drivers/misc/xlink-pcie/remote_host/pci.h b/drivers/misc/xlink-pcie/remote_host/pci.h
index 09499b50fd0e..941b01fc2c63 100644
--- a/drivers/misc/xlink-pcie/remote_host/pci.h
+++ b/drivers/misc/xlink-pcie/remote_host/pci.h
@@ -44,6 +44,11 @@ struct xpcie_dev {
 	struct work_struct irq_event;
 	bool boot_dev_link;
 	mxlk_pcie_boot_event boot_notif_fn;
+
+	/* FlashLogic DMA allocation */
+	dma_addr_t fl_phys_addr;
+	void *fl_vbuf;
+	size_t fl_buf_size;
 };
 
 static inline struct device *xpcie_to_dev(struct xpcie *xpcie)
-- 
2.25.1

